### PR TITLE
make IntervalOptimisation and SOS stuff optional

### LINKED ISSUE #40: Make SDPA optional

### PR DESCRIPTION
**Pull request by lucaferranti**
- make IntervalOptimisation and SumOfSquares optional dependencies.
- TaylorModels also should be optional I think, but it seems to be hardcoded in the branchandbound atm, so I'd leave for a separate PR
- similar for DynamicPolynomials
- good to notice that SDPA doesn't work anyway on 1.7 hence not tested
- fixes #40

### CODE REVIEW

**Comment by lucaferranti** [comment]
@schillic @mforets I cannot add you as reviewers for some reasons so here a good old fashioned ping

File: src/intervaloptimisation.jl [hunk 1]
```diff
@@ -0,0 +1,98 @@
+#===================================
+Methods using Interval Optimisation
+===================================#
+using .IntervalOptimisation
+
+"""
+    range_and_optimisers(f::Function, dom::Interval_or_IntervalBox;
+                         structure=HeapedVector, tol=1e-3)
```
Comment (schillic) [inline src/intervaloptimisation.jl:8]: In the docs build:

```julia
ERROR: UndefVarError: HeapedVector not defined
```

Could be related to the some dependency now being optional and hence not loaded in the docs build.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,7 +1,9 @@
-using Test, RangeEnclosures
-using RangeEnclosures: relative_precision, available_solvers
+using IntervalOptimisation, Test, RangeEnclosures
+using RangeEnclosures: relative_precision
 
 using DynamicPolynomials: @polyvar
 
+available_solvers = (:IntervalArithmetic, :TaylorModels, :BranchandBound, :IntervalOptimisation)
```
Comment (schillic) [inline test/runtests.jl:6]: Why not use `available_solvers` like before?

File: test/univariate.jl [hunk 3]
```diff
@@ -38,10 +38,10 @@ end
...
     @test inf(r) ≤ 1e-5 && sup(r) ≤ 1e-5
 
-    x = enclose(f, dom, :SumOfSquares)
-    xref = Interval(4.8333, 10.541)
-    r = relative_precision(x, xref)
-    @test inf(r) ≤ 1e-5 && sup(r) ≤ 1e-5
+    # x = enclose(f, dom, :SumOfSquares)
+    # xref = Interval(4.8333, 10.541)
+    # r = relative_precision(x, xref)
+    # @test inf(r) ≤ 1e-5 && sup(r) ≤ 1e-5
```
Comment (schillic) [inline test/univariate.jl:44]: Why is that needed?

File: src/intervaloptimisation.jl (hunk 1 above)
Comment (lucaferranti) [reply src/intervaloptimisation.jl:8]: yes I think this is the case, I think adding IntervalOptimisation.jl to the docs Project.toml would fix it?

File: test/univariate.jl (hunk 3 above)
Comment (schillic) [reply test/univariate.jl:44]: It does, only not on v1.7.

Comment (schillic) [reply test/univariate.jl:44]: So I suggest to guard these tests in a `if VERSION < v"1.7"` block. But we can also handle these tests in another PR when testing on v1.7.

**Comment by lucaferranti** [comment]
```julia
ERROR: LoadError: `makedocs` encountered an error. Terminating build
```

thanks for the informative error message

File: test/univariate.jl (hunk 3 above)
Comment (schillic) [reply test/univariate.jl:44]: I am not sure if there are Windows-specific problems. v1.7 also does not work on Linux and v1.6 works on Windows.

**Comment by schillic** [comment]
These are new warnings in the docs build. Maybe they help in debugging.

```julia
┌ Warning: undefined binding 'RangeEnclosures.enclose_SumOfSquares' in `@docs` block in src/lib/methods.md:28-33
│ ```@docs
│ enclose_IntervalArithmetic
│ enclose_IntervalOptimisation
│ enclose_TaylorModels
│ enclose_SumOfSquares
│ ```
└ @ Documenter.Expanders ~/.julia/packages/Documenter/PLD7m/src/Expanders.jl:298
┌ Warning: undefined binding 'RangeEnclosures.new_sos' in `@docs` block in src/lib/methods.md:43-45
│ ```@docs
│ new_sos
│ ```
└ @ Documenter.Expanders ~/.julia/packages/Documenter/PLD7m/src/Expanders.jl:298
```

File: test/runtests.jl (hunk 2 above)
Comment (schillic) [reply test/runtests.jl:6]: @mforets do you remember?

**Comment by schillic** [comment]
The last commit fixed the docs build but failed the v1.0 build :sweat_smile:

**Comment by lucaferranti** [comment]
it didn't but that seems indeed to be the reason, compare the current log for 1.0 to [this one](https://github.com/JuliaReach/RangeEnclosures.jl/runs/5687806236?check_suite_focus=true), now e.g. Mutable Arithmetic is locked to 0.3.3 while on the old log it was at 0.2.22

**Comment by schillic** [comment]
Yes, I think there are two options:
1. Bump the lower bound such that the tests work.
2. Use a custom build script that installs the old package version before the build step.

**Comment by mforets** [comment]
No need to test in julia v1.0, can you remove it from CI?

File: test/runtests.jl (hunk 2 above)
Comment (mforets) [reply test/runtests.jl:6]: Ideally we shouldn't need such tuple at all, and just rely on dispatch?

Let's focus on getting the compilation issue right on this PR.

Comment (lucaferranti) [reply test/runtests.jl:6]: I mean, even if you use dispatch, you probably still want a tuple `solvers_to_test = (Solver1(), Solver2(), Solver3())` to test the `enclose` function? Note this is in the `runtest.jl` file. I agree that such a tuple is not needed inside the source code (indeed atm as far as I can say even now it's not used), but this can be delegated to future RP

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -21,11 +21,8 @@ jobs:
       matrix:
         julia-version: ['1.6']
         julia-arch: [x64]
-        os: [ubuntu-latest, windows-latest]
-        include:
-          - julia-version: '1.0'  # test on oldest supported version
```
Comment (schillic) [inline .github/workflows/ci.yml:26]: I suggest:

```julia
julia-version: ['1.7']
        julia-arch: [x64]
        os: [ubuntu-latest, windows-latest]
        include:
          - julia-version: '1.6'  # test on oldest supported version
```

File: docs/Project.toml [hunk 5]
```diff
@@ -1,5 +1,11 @@
...
 Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
+IntervalOptimisation = "c7c68f13-a4a2-5b9a-b424-07d005f8d9d2"
+SDPA = "b9a10b5b-afa4-512f-a053-bb3d8080febc"
+SumOfSquares = "4b9e565b-77fc-50a5-a571-1244f986bda1"
 
 [compat]
 Documenter = "0.24"
+IntervalOptimisation = "0.4.1"
+SDPA = "0.2, 0.3, 0.4"
+SumOfSquares = "0.3.6, 0.4, 0.5, 0.6"
```
Comment (schillic) [inline docs/Project.toml:11]:
```suggestion
SumOfSquares = "0.3.6, 0.4, 0.5, 0.6"

```

File: src/RangeEnclosures.jl [hunk 6]
```diff
@@ -1,30 +1,33 @@
...
                            :TaylorModels,
-                           :SumOfSquares,
                            :BranchandBound]
 
-const optional_solvers = [:AffineArithmetic]
+const optional_solvers = [:AffineArithmetic, :IntervalOptimisation, :SumOfSquares]
 
 include("intervals.jl")
 include("taylormodels.jl")
-include("sdp.jl")
```
Comment (schillic) [inline src/RangeEnclosures.jl:20]: Why did you remove this include?

File: docs/Project.toml (hunk 5 above)
Comment (lucaferranti) [reply docs/Project.toml:11]: what's the difference?

File: src/RangeEnclosures.jl (hunk 6 above)
Comment (lucaferranti) [reply src/RangeEnclosures.jl:20]: it depends on SumOfSquares.jl which is an optional dependency, it's imported in the init function

File: docs/Project.toml (hunk 5 above)
Comment (lucaferranti) [reply docs/Project.toml:11]: aah empty line at the end

File: .github/workflows/ci.yml (hunk 4 above)
Comment (lucaferranti) [reply .github/workflows/ci.yml:26]: should it be '1' instead of '1.7' ? it would automatically mean "latest stable release". Also should 1.6 be tested on all OS ?

Comment (schillic) [reply .github/workflows/ci.yml:26]: We do it like that in all packages. No strong opinion, but deciding which version is used has its merits.

Usually we only test the lower bound on one system. It is more a warning system that we do not accidentally drop support for some feature. To really do proper testing we would also have to use the lowest package versions.

**Comment by schillic** [comment]
```julia
Failed to add the RSA host key for IP address '140.82.114.3' to the list of known hosts (/home/runner/.ssh/known_hosts).
Load key "/home/runner/work/RangeEnclosures.jl/RangeEnclosures.jl/docs/.documenter": invalid format
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
┌ Error: Git failed to fetch git@github.com:JuliaReach/RangeEnclosures.jl.git
│ This can be caused by a DOCUMENTER_KEY variable that is not correctly set up.
│ Make sure that the environment variable is properly set up as a Base64-encoded string
│ of the SSH private key. You may need to re-generate the keys with DocumenterTools.
```
Docs previews worked before, but I do not see why this PR would break them. Maybe because it comes from a fork.

**Comment by lucaferranti** [comment]
I think so. IIRC I had the same issue when Jorge opened a PR from his fork to ILA

**Review by schillic — APPROVED** [review]

**Comment by mforets** [comment]
> Docs previews worked before, but I do not see why this PR would break them. Maybe because it comes from a fork.

That seems right.

**Review by mforets — APPROVED** [review]

**Comment by schillic** [comment]
Yes, docs deployment worked on `master`.

File: test/runtests.jl (hunk 2 above)
Comment (mforets) [reply test/runtests.jl:6]: > I mean, even if you use dispatch, you probably still want a tuple solvers_to_test = (Solver1(), Solver2(), Solver3()) to test the enclose function? 

Correct; the idea here was to iterate over some specified set of solver backends for the same problem.
